prov/hook/hook_ze/src/hook_ze.c

```diff
@@ -0,0 +1,1887 @@
+		const struct iovec *iov, struct hook_ze_desc **ze_desc)
+{
+	struct fi_mr_attr attr = {0};
+	struct iovec base_iov;
+	uint64_t iface, device, flags;
+	int ret = 0;
+
+	*ze_desc = ofi_buf_alloc(domain->mr_pool);
+	if (!*ze_desc)
+		return -FI_ENOMEM;
+
+	iface = ze_is_addr_valid(iov->iov_base, &device,
+				 &flags) ? FI_HMEM_ZE : FI_HMEM_SYSTEM;
+
+	if (iface != FI_HMEM_ZE) {
+		(*ze_desc)->desc = NULL;
+		(*ze_desc)->iov = *iov;
+		goto out;
+	}
+
+	ret = ze_hmem_get_base_addr(iov->iov_base, &base_iov.iov_base,
+				    &base_iov.iov_len);
+	if (ret)
+		return -FI_EINVAL;
+
+	attr.mr_iov = &base_iov;
+	attr.iov_count = 1;
+	attr.access = FI_SEND | FI_RECV | FI_READ | FI_WRITE |
+		      FI_REMOTE_READ | FI_REMOTE_WRITE;
+	attr.offset = 0;
+	attr.requested_key = (uint64_t) *ze_desc;
+	attr.iface = FI_HMEM_ZE;
+	attr.device.ze = (int) device;
+
+	ret = fi_mr_regattr(domain->hook_domain.hdomain, &attr,
+			    (*ze_desc)->flags, &(*ze_desc)->mr_fid);
+	if (ret) {
+		ofi_buf_free(*ze_desc);
+		return ret;
+	}
+
+	(*ze_desc)->desc = fi_mr_desc((*ze_desc)->mr_fid);
+	(*ze_desc)->iov = base_iov;
+out:
+	(*ze_desc)->count = 0;
+	dlist_insert_tail(&(*ze_desc)->entry, &domain->mr_list);
+	return FI_SUCCESS;
+}
+
+static int hook_ze_cache_mr(struct hook_ze_domain *domain,
+			    const struct iovec *iov,
+			    struct hook_ze_desc **ze_desc)
+{
+	struct ofi_rbnode *node;
+	int ret;
+
+	ret = ofi_rbmap_insert(&domain->rbmap, (void *) iov, NULL, &node);
+	if (!ret) {
+		ret = hook_ze_add_region(domain, iov,
+				(struct hook_ze_desc **) (&node->data));
+		if (ret) {
+			ofi_rbmap_delete(&domain->rbmap, node);
+			return ret;
+		}
+	} else if (ret != -FI_EALREADY) {
+		return ret;
+	}
+
+	*ze_desc = (struct hook_ze_desc *) (node->data);
+	(*ze_desc)->count++;
+
+	return FI_SUCCESS;
+}
+
+static void hook_ze_uncache_mr(struct hook_ze_domain *domain,
+			       const struct iovec *iov)
+{
+	struct ofi_rbnode *node;
+	struct hook_ze_desc *ze_desc;
+
+	node = ofi_rbmap_find(&domain->rbmap, (void *) iov);
+	if (!node)
+		return;
+
+	ze_desc = (struct hook_ze_desc *) (node->data);
+	if (--ze_desc->count)
+		return;
+
+	ofi_rbmap_delete(&domain->rbmap, node);
+	if (ze_desc->desc)
+		fi_close(&(ze_desc->mr_fid)->fid);
+
+	dlist_remove(&ze_desc->entry);
+	ofi_buf_free(ze_desc);
+}
+
+static void hook_ze_uncache_mr_iov(struct hook_ze_domain *domain,
+				   const struct iovec *iov, size_t count)
+{
+	int i;
+
+	for (i = 0; i < count; i++)
+		hook_ze_uncache_mr(domain, &iov[i]);
+}
+
+static int hook_ze_cache_mr_iov(struct hook_ze_domain *domain,
+				const struct iovec *iov, void **desc,
+				size_t count, struct hook_ze_desc **ze_desc)
+{
+	int i, ret;
+
+	for (i = 0; i < count; i++) {
+		ret = hook_ze_cache_mr(domain, &iov[i], &ze_desc[i]);
+		if (ret) {
+			hook_ze_uncache_mr_iov(domain, iov, i);
+			return ret;
+		}
+
+		desc[i] = ze_desc[i]->desc ? ze_desc[i]->desc : desc[i];
+	}
+	return FI_SUCCESS;
+}
```

Comment:
Can the existing mr_cache implementation be used instead of re-implementation portions of it?

Reply:
I originally wrote it with the MR cache code and then I changed it for some reason that I cannot think of and then I reworked the entire hook so it could very well be that it could go back in. I'll take a look and try to remember why I changed it....

Reply:
I didn't do a comparison of the code.  It could be the hook_ze_add_region() call that you need to made that doesn't work.

Reply:
@shefty Ok I remember now. The MR cache implementation requires the iface when searching/registering memory which the hook doesn't have when it searches the map since it's trying to figure that part out. The hook usage is actually closer to the mr_map but that's based on keys/access
So there could be a few options:
- modify the mr_map to take an add/delete/compare function to specialize it
- modify/extend the mr_cache to support this use case (not sure the impact here)
- keep this usage separate (potentially adding the memory monitor here - how important is the memory monitor here?)